src/httpserver.rs

```diff
@@ -1,178 +1,228 @@
-use std::{convert::Infallible, sync::Arc, time::Instant};
+use std::{
+    collections::HashMap,
+    error::Error,
+    io::{BufRead, BufReader, Read, Write},
+    net::{Shutdown, TcpListener, TcpStream},
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+    thread,
+    time::{Duration, Instant},
+};
 
 use hmac::{Hmac, Mac};
-use hyper::service::{make_service_fn, service_fn};
-use hyper::{self, body::Bytes, server::conn::AddrIncoming, Body, Request, Response, StatusCode};
 use percent_encoding::percent_decode;
 use secstr::SecStr;
-use sha1::Sha1;
+use sha2::Sha256;
 
 use crate::{queue::QueueJob, Snare};
 
-pub(crate) async fn serve(server: hyper::server::Builder<AddrIncoming>, snare: Arc<Snare>) {
-    let make_svc = make_service_fn(|_| {
+/// How many connections to accept simultaneously? Limiting this number stops attackers from
+/// causing us to use too many resources.
+static MAX_SIMULTANEOUS_CONNECTIONS: usize = 16;
+/// How long to try reading/writing from a socket before we assume it's died.
+static NET_TIMEOUT: Duration = Duration::from_secs(10);
+/// The maximum payload size we'll accept from a remote in bytes. The main reason to limit this is
+/// to stop large numbers of requests causing us to run out of memory.
+static MAX_HTTP_BODY_SIZE: usize = 64 * 1024;
+
+pub(crate) fn serve(snare: Arc<Snare>) -> Result<(), Box<dyn Error>> {
+    let listener = TcpListener::bind(snare.conf.lock().unwrap().listen)?;
+    let active = Arc::new(AtomicUsize::new(0));
+    for mut stream in listener.incoming().flatten() {
+        while active.fetch_add(1, Ordering::Relaxed) > MAX_SIMULTANEOUS_CONNECTIONS {
```

Comment:
While adding one is greater than what we are permitting to accept, subtract one... This is clever in a way I'm not sure what it does, and needs a comment I reckon!

Reply:
Hopefully addressed in 36f8f5e.

Reply:
I think I get it.